src/errors.rs

```diff
@@ -0,0 +1,19 @@
+use thiserror::Error;
+
+/// Generic input/output error with context message
+#[derive(Error, Debug)]
+#[error("{msg}. Cause: {source}")]
+pub struct IoError {
+    source: std::io::Error,
+    msg: String,
+}
+
+impl IoError {
+    /// Constructs an IoError from an std IO error and a context message
+    pub fn new(source: std::io::Error, msg: impl Into<String>) -> Self {
```

Comment:
should this be beyond `pub(crate)`?

Reply:
I need to give this one more thought. The goal of `thiserror` is to move all the static details of an Error type to its struct declaration. That means this struct goes against `thiserror`'s philosophy, as it captures the context inline and makes it opaque to the wrapper struct. So when I wrote this I wasn't sure if I was doing a good thing.

That said, looking at this from a consumer's perspective, they probably don't care exactly _how_ an IO error was produced, just that it's an IO error, and we're capturing the context in the message field. So maybe I was actually on the right track and should do more of this.

Reply:
I think we should have `thiserror` enums and capture them in an `eyre::Report`-like structure which can keep track of traces. So i imagine there can be multiple errors which affect eachother that should be stacked / traced. Eg. when trying to install a dependency and it fails due to missing credentials i imagine something like this:

```
1. InstallationError
2. AuthenticationError
3. CredentialsError```

I dont know whether using an eyre report or a custom impl is better – the benefit of implementing this would be the detailed control over how it is printed in the end + the typesafety